### PR TITLE
Add benchmark suite

### PR DESCRIPTION
**Pull request by 0ortmann**
###  :notebook_with_decorative_cover: Description

Add a (very simple) benchmark suite.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/threatbus](https://docs.tenzir.com/threatbus), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

The `benchmark/README` contains all instructions for running a benchmark. Same instructions can be used to  test this PR

### CODE REVIEW

**Comment by mavam** [comment]
A small diff when going through the README so far:

```diff
diff --git a/benchmark/README.md b/benchmark/README.md
index 6349205..933409c 100644
--- a/benchmark/README.md
+++ b/benchmark/README.md
@@ -2,9 +2,9 @@ Benchmarking
 ============

 This folder provides experimental benchmarking tools to measure message passing
-performance for some Threat Bus plugins.
+performance for Threat Bus plugins.

-Benchmarks are executed via the script `simple-bench`. All benchmarks must be
+The script `simple-bench` executes the benchmarks. All benchmarks must be
 executed from within the `benchmark` folder, because subsequent invocations are
 performed relatively to the current working directory of the caller.

@@ -19,6 +19,7 @@ virtualenv venv
 source venv/bin/activate

 # main threatbus
+cd ..
 python setup.py develop

 # zmq-app plugin
@@ -28,7 +29,7 @@ python plugins/apps/threatbus_zmq_app/setup.py develop
 python plugins/backbones/file_benchmark/setup.py develop

 # rabbit backbone
-python plugins/backbones/rabbitmq/setup.py develop
+python plugins/backbones/threatbus_rabbitmq/setup.py develop

 # start benchmarking
 cd benchmark
@@ -37,7 +38,7 @@ cd benchmark

 Start benchmarking when you have installed all the above packages. Invoke the
 `simple-bench.sh` script always from within the `benchmark` folder. The script
-takes four arguments:
+takes 4 arguments:

 1. The type of the benchmark (either `rabbit_consumer` or `zmq`)
 2. The number of rounds to run
```

The preparation commands ran through. But the benchmark itself not:

```
./simple-bench.sh rabbit_consumer 2 100 3
Benchmarking RabbitMQ consumer
Calculating the difference between first and last message processed, in milliseconds:
rm: threatbus.log: No such file or directory
./simple-bench.sh: line 8: ../venv/bin/threatbus: No such file or directory
grep: threatbus.log: No such file or directory
Fewer messages logged than sent! Try increasing the sleep interval:        0/100
grep: threatbus.log: No such file or directory
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
rm: threatbus.log: No such file or directory
./simple-bench.sh: line 8: ../venv/bin/threatbus: No such file or directory
grep: threatbus.log: No such file or directory
Fewer messages logged than sent! Try increasing the sleep interval:        0/100
grep: threatbus.log: No such file or directory
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
```

I pushed a simple fix that gets rid of the warning in the benchmark script. I'm not going to do further major fixups directly, but this patch allowed me to run Threat Bus:

```diff
diff --git a/benchmark/simple-bench.sh b/benchmark/simple-bench.sh
index 941ff2b..e0390c7 100755
--- a/benchmark/simple-bench.sh
+++ b/benchmark/simple-bench.sh
@@ -5,7 +5,7 @@ rabbit_consumer() {
   rm -f threatbus.log
   python rabbitmq_sender.py $1

-  ../venv/bin/threatbus -c benchmark_config.yaml &
+  threatbus -c benchmark_config.yaml &
   sleep $2
   count=$(grep 'Relayed message from RabbitMQ' threatbus.log | wc -l)
   if [ $count != $1 ]; then
@@ -20,7 +20,7 @@ zmq() {
   # Starts Threat Bus and sends messages to `zmq-app` plugin endpoint. Starts a dummy consumer. Measures roundtrip of receiving
   rm -f threatbus.log
   sed -i "s/repetitions.*/repetitions: $1/" benchmark_config.yaml
-  ../venv/bin/threatbus -c benchmark_config.yaml &
+  threatbus -c benchmark_config.yaml &
   sleep 0.5
   python ../tests/utils/zmq_receiver.py $1 1>/dev/null &
   sleep $2
```

The relative path is not neccessary because we are already getting $PATH from the venv, no?

**Comment by mavam** [comment]
After my benchmark script fix, I'm still getting errors though:

```
  File "/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/mavam/code/tenzir/threatbus/plugins/apps/threatbus_cif3/plugin.py", line 3, in <module>
    from cifsdk.client.http import HTTP as Client
ModuleNotFoundError: No module named 'cifsdk'
```

**Comment by 0ortmann** [comment]
Yes, right, we get $PATH from `venv`. I'll change that. For the `cifsdk` - the error message tells you that it's not installed. Did you run either `make dev-mode` or `make install`, @mavam?

**Comment by 0ortmann** [comment]
I updated this branch to reflect and work with the latest changes from master.

**Comment by 0ortmann** [comment]
@mavam mind giving this another spin? would be great to release with a clean slate.

**Comment by mavam** [comment]
Yes, I'll take a look now.

File: benchmark/simple-bench.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env sh
+
+rabbit_consumer() {
+  # Injects messages into RabbitMQ, then starts Threat Bus and measures raw consumption times.
+  rm threatbus.log
```
Comment (mavam) [inline benchmark/simple-bench.sh:5]:
```suggestion
  rm -f threatbus.log
```

Do not display non-existant file warnings.

File: benchmark/simple-bench.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    echo "Fewer messages logged than sent! Try increasing the sleep interval: $count/$1"
+  fi
+  grep 'Relayed message from RabbitMQ' threatbus.log | sed -n '1p;$p' | awk -F ' ' '{print $2}' | awk 'NR > 1 {"date -d "$0" +%s%N"|getline a; "date -d "prev" +%s%N"|getline b; print (a-b)/1000000} {prev=$0}'
+
+  kill $(pgrep threatbus)
+}
+
+zmq() {
+  # Starts Threat Bus and sends messages to `zmq-app` plugin endpoint. Starts a dummy consumer. Measures roundtrip of receiving and publishing messages.
+  rm threatbus.log
```
Comment (mavam) [inline benchmark/simple-bench.sh:21]:
```suggestion
  rm -f threatbus.log
```

**Review by mavam — APPROVED** [review]
This appears to work now:

```
./simple-bench.sh rabbit_consumer 2 100 3 2> /dev/null
Benchmarking RabbitMQ consumer
Calculating the difference between first and last message processed, in milliseconds:
26.0001
24.9999
```

And:

```
./simple-bench.sh zmq 3 10 3
Benchmarking ZMQ consumer
Calculating the difference between first and last message processed, in milliseconds:
1.99987
./simple-bench.sh: line 35: 39495 Terminated: 15          threatbus -c "$config"
/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 32 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
4.99994
./simple-bench.sh: line 35: 39516 Terminated: 15          threatbus -c "$config"
/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 32 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
1.99987
/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 32 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

Note the warnings when not ignoring stderr. However, I'm fine with this potentially OS-specific glitch.

**Pull request merged**